<img src='https://user-images.githubusercontent.com/6457691/89746244-a6e2a200-daf3-11ea-9314-7863d544d73c.png' width = '200' align = 'right'>

## *DATA SCIENCE / SECTION 1 / SPRINT 1 / NOTE 2*

# 📝 Assignment


## 1. Feature Engineering

- `KT&G` 기업 데이터를 계속 사용합니다.
  - 변수 이름은 `df`로 저장하세요.

- 각 분기별 영업이익률을 직접 계산해보세요. 
  - Feature 이름은 `영업이익률2` 입니다.

- 아래 데이터를 참조하여 18/19년도에 대해서 Price to Dream Ratio (PDR)을 계산하세요.
  - 변수 이름은 `df2`, feature 이름은 `PDR`로, `df2['PDR']`을 통해 `~` 값이 확인 되어야합니다. 아래의 전제조건을 활용하세요.
  - 점유율 (일반담배 기준) 
    - 18년 : 62%
    - 19년 : 63.5% 
  - 시장의 규모 
     - 18년 : 15조 
     - 19년 : 15조

||구분|종가|발행주식수|시가총액|시장점유율|PDR|
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|0|18년|101500|137292497|13935188445500|62.0|~|
|1|19년|93800|137292497|12878036218600|63.5|~|


In [ ]:
url = 'https://ds-lecture-data.s3.ap-northeast-2.amazonaws.com/kt%26g/kt%26g_0.csv'
headers = ['분기', '매출액', '영업이익', '영업이익(발표기준)', '세전계속사업이익',
           '당기순이익', '당기순이익(지배)', '당기순이익(비지배)', '자산총계', '부채총계',
           '자본총계', '자본총계(지배)', '자본총계(비지배)', '자본금', '영업활동현금흐름',
           '투자활동현금흐름', '재무활동현금흐름', '영업이익률', '순이익률', 'ROE(%)',
           'ROA(%)', '부채비율', '자본유보율', 'EPS(원)', 'PER(배)']



In [ ]:
# header 넣어줌
import pandas as pd

df = pd.read_csv(url,header= None) # 원래 있던 데이터가 header가 되어 사라지지 않게 header=None 추가

df.columns = headers # 위에 지정한 header를 column(header)로 
df

분기     매출액   영업이익 영업이익(발표기준)  ...   부채비율   자본유보율 EPS(원)  PER(배)
0   19/6  12,578  3,991      3,991  ...  31.43  803.22  2,351   13.77
1   19/9  13,222  3,825      3,825  ...  24.17  836.86  2,337   13.87
2  19/12  11,982  2,523      2,523  ...  23.04  850.82    873   12.42
3   20/3  11,784  3,150      3,150  ...  29.69  823.16  2,134    9.71
4   20/6  13,188  3,947      3,947  ...  26.91  837.40  2,135   10.46

[5 rows x 25 columns]

In [ ]:
# 분기별 영업이익률 _ 영업이익률2

# 영업이익률: 영업이익/매출액

def toInt(string):
  return int(string.replace(',',''))


df['영업이익'] = df['영업이익'].apply(toInt)
df['매출액'] = df['매출액'].apply(toInt)

# 영업이익률2

df['영업이익률2'] = round((df['영업이익']/df['매출액'])*100,2) #소수점 2자리 이하 반올림

df
df.dtypes


분기             object
매출액             int64
영업이익            int64
영업이익(발표기준)     object
세전계속사업이익       object
당기순이익          object
당기순이익(지배)      object
당기순이익(비지배)      int64
자산총계           object
부채총계           object
자본총계           object
자본총계(지배)       object
자본총계(비지배)     float64
자본금            object
영업활동현금흐름       object
투자활동현금흐름       object
재무활동현금흐름       object
영업이익률         float64
순이익률          float64
ROE(%)        float64
ROA(%)        float64
부채비율          float64
자본유보율         float64
EPS(원)         object
PER(배)        float64
영업이익률2        float64
dtype: object

In [ ]:
# 18/19 년도 PDR 계산

# PDR : 시가총액 / (전체 시장 규모(TAM) * (시장점유율/100)) * 100
df2 = pd.DataFrame({'구분':['18년','19년'], 
                    '종가':[101500,93800], 
                    '발행주식수':[137292497,137292497],
                    '시가총액':[13935188445500,12878036218600],
                    '시장점유율':[62.0,63.5]})                    # DataFrame으로 만들어주기
df2['PDR'] = df2['시가총액']/(15000000000000*(df2['시장점유율']/100))*100 # 'PDR' 추가
df2


구분      종가      발행주식수            시가총액  시장점유율         PDR
0  18년  101500  137292497  13935188445500   62.0  149.840736
1  19년   93800  137292497  12878036218600   63.5  135.202480

## 2. String Manipulation

- replace가 아닌 다른 방법을 사용하여 19년도 2분기의 매출액 을 integer로 변경하세요. 결과값은 12578이 나와야합니다.
  


In [ ]:
# 19년도 2분기 매출액 integer로 변경

df['매출액'] = df['매출액'].astype(int)

df['매출액'][0]

12578

## 3. Apply
- 데이터셋의 분기를 제외한 모든 feature에 대해서 수치형 데이터 (`int`, `float`) 로 수정하고, `describe`를 통해 summary statistics 를 확인하세요.
- 단 결측치는 `0`으로 처리합니다.

In [ ]:
# 데이터셋의 분기를 제외한 모든 feature에 대해서 수치형 데이터 (int, float) 로 수정하고, describe를 통해 summary statistics 를 확인

#regex=True : delete all comma
df_edit = df.columns[1:] # 제일 처음 column 제거

df[df_edit] = df[df_edit].replace(',','',regex= True).astype(float)
df.dtypes

# 결측치 0으로 대치
df.fillna(0)
df.describe()



매출액         영업이익  ...     PER(배)     영업이익률2
count      5.000000     5.000000  ...   5.000000   5.000000
mean   12550.800000  3487.200000  ...  12.046000  27.676000
std      664.979097   636.931079  ...   1.898112   4.115651
min    11784.000000  2523.000000  ...   9.710000  21.060000
25%    11982.000000  3150.000000  ...  10.460000  26.730000
50%    12578.000000  3825.000000  ...  12.420000  28.930000
75%    13188.000000  3947.000000  ...  13.770000  29.930000
max    13222.000000  3991.000000  ...  13.870000  31.730000

[8 rows x 25 columns]

## 🔥 도전과제



### NA Value Handling

- 19년도 4분기의 당기순이익(비지배) 부분을 Na로 대체하세요
- 이후 해당 결측치를 mean imputation 방법을 사용하여 처리하세요.



In [ ]:
# 19년도 4분기의 당기순이익(비지배) 부분을 Na로 대체

df.iloc[2,7] = 'NaN'

# 해당 결측치를 mean imputation 방법을 사용하여 처리

import numpy as np
from sklearn.impute import SimpleImputer

#np는 NaN을 처리해서 Nan 으로 바꿔줌! 
imputer = SimpleImputer(missing_values=np.nan,strategy = 'mean')
answer = imputer.fit_transform(df[['당기순이익(비지배)','자본총계(비지배)']])
df[['당기순이익(비지배)','자본총계(비지배)']] = answer
df[['당기순이익(비지배)','자본총계(비지배)']].isnull().sum()

df

분기      매출액    영업이익  영업이익(발표기준)  ...   자본유보율  EPS(원)  PER(배)  영업이익률2
0   19/6  12578.0  3991.0      3991.0  ...  803.22  2351.0   13.77   31.73
1   19/9  13222.0  3825.0      3825.0  ...  836.86  2337.0   13.87   28.93
2  19/12  11982.0  2523.0      2523.0  ...  850.82   873.0   12.42   21.06
3   20/3  11784.0  3150.0      3150.0  ...  823.16  2134.0    9.71   26.73
4   20/6  13188.0  3947.0      3947.0  ...  837.40  2135.0   10.46   29.93

[5 rows x 26 columns]

### Feature Engineering

Relative Perfomance 라는 새로운 feature를 계산하세요.

이는 최근 1년치 매출액의 평균값을 기준으로 

- 10% 이상 -> `S`
- 5% 이상 -> `A`
- -5 ~ 5% -> `B`
- -5%이하 -> `C`
- 10%이하 -> `D`

라는 값을 갖는 feature입니다.

20년도 2분기에 해당하는 결과값은 `A`가 나와야합니다.

각각에 해당하는 등급이 나오기 위해서 필요한 매출액을 추가로 서술하세요.

In [ ]:
# Relative Perfomance 라는 새로운 feature를 계산

# 4분기의 평균 계산

mean2 = df.iloc[1:, 1].sum() / 4
print('mean2:',mean2)

# 점수 계산 : 평균매출액 * 110%, 105%, 105-95%, 90%
for idx, i in enumerate(df['매출액']):
  print(i,idx)
  if i >= (mean2*1.1):
    df['Relative Performance'][idx] = 'S'
  elif i>= (mean2*1.05):
    df['Relative Performance'][idx] = 'A'
  elif i>= (mean2*0.95):
    df['Relative Performance'][idx] = 'B'
  elif i>= (mean2*0.9):
    df['Relative Performance'][idx] = 'C'
  else:
    df['Relative Performance'][idx] = 'D'

# 각각에 해당하는 등급이 나오기 위해 필요한 매출액

print('S :', round(mean2*1.1,2))
print('A :', round(mean2*1.05,2))
print('B :', round(mean2*0.95,2))
print('C :', round(mean2*0.9,2))
print('D :', round(mean2*0.9,2))

df


mean2: 12544.0
12578.0 0
13222.0 1
11982.0 2
11784.0 3
13188.0 4
S : 13798.4
A : 13171.2
B : 11916.8
C : 11289.6
D : 11289.6


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  app.launch_new_instance()
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


분기      매출액    영업이익  ...  PER(배)  영업이익률2  Relative Performance
0   19/6  12578.0  3991.0  ...   13.77   31.73                     B
1   19/9  13222.0  3825.0  ...   13.87   28.93                     A
2  19/12  11982.0  2523.0  ...   12.42   21.06                     B
3   20/3  11784.0  3150.0  ...    9.71   26.73                     C
4   20/6  13188.0  3947.0  ...   10.46   29.93                     A

[5 rows x 27 columns]